apollo-federation/src/composition/satisfiability/satisfiability_error.rs

```diff
@@ -0,0 +1,466 @@
+                        bail!("generate_witness_value: enum type has no values");
+                    };
+                    ast::Value::Enum(first_value.clone())
+                }
+                TypeDefinitionPosition::InputObject(input_object_type_pos) => {
+                    let object_type = input_object_type_pos.get(schema.schema())?;
+                    let fields = object_type
+                        .fields
+                        .iter()
+                        .filter_map(|(field_name, field_def)| {
+                            // We don't bother with non-mandatory fields.
+                            if !field_def.is_required() {
+                                return None;
+                            }
+
+                            let field_value = match generate_witness_value(schema, field_def) {
+                                Ok(value) => value,
+                                Err(e) => {
+                                    return Some(Err(e));
+                                }
+                            };
+                            Some(Ok((field_name.clone(), field_value)))
+                        })
+                        .collect::<Result<Vec<_>, _>>()?;
+                    ast::Value::Object(fields)
+                }
+                _ => bail!("generate_witness_value: unexpected value type"),
+            }
+        }
+        executable::Type::List(_item_type) | executable::Type::NonNullList(_item_type) => {
+            ast::Value::List(vec![])
+        }
+    };
+    Ok(Node::new(value))
+}
+
+fn display_reasons(reasons: &[Unadvanceables]) -> String {
+    let mut by_subgraph = MultiMap::new();
+    for reason in reasons {
+        for unadvanceable in reason.iter() {
+            by_subgraph.insert(unadvanceable.source_subgraph(), unadvanceable)
+        }
+    }
+    by_subgraph
+        .iter_all()
+        .filter_map(|(subgraph, reasons)| {
+            let (first, rest) = reasons.split_first()?;
+            let details = if rest.is_empty() {
+                format!(r#" {}."#, first.details())
+            } else {
+                // We put all the reasons into a set because it's possible multiple paths of the
+                // algorithm had the same "dead end". Typically, without this, there is cases where we
+                // end up with multiple "cannot find field x" messages (for the same "x").
+                let all_details = reasons
+                    .iter()
+                    .map(|reason| reason.details())
+                    .collect::<IndexSet<_>>();
+                let mut formatted_details = vec!["".to_string()]; // to add a newline
+                formatted_details
+                    .extend(all_details.iter().map(|details| format!(" - {details}.")));
+                formatted_details.join("\n")
+            };
+            Some(format!(r#"- from subgraph "{subgraph}": {details}"#))
```

Comment:
nit: To match JS spacing, this should be
```suggestion
            Some(format!(r#"- from subgraph "{subgraph}":{details}"#))
```